pkg/builtins/ec_op.go

```
@@ -0,0 +1,262 @@
+	}
+
+	alpha_big_int := big.NewInt(1)
+
+	partial_sum := PartialSum{X: input_cells[0], Y: input_cells[1]}
+	double_point := DoublePoint{X: input_cells[2], Y: input_cells[3]}
+
+	result, err := EcOnImpl(partial_sum, double_point, input_cells[4], alpha_big_int, prime, ec.scalar_height)
+
+	felt_result_x := lambdaworks.FeltFromHex(result.X.Text(16))
+	felt_result_y := lambdaworks.FeltFromHex(result.Y.Text(16))
+
+	ec.cache[x_addr] = felt_result_x
+	ec.cache[x_addr.AddUint(1)] = felt_result_y
+
+	if index-uint(ec.n_input_cells) == 0 {
+		return memory.NewMaybeRelocatableFelt(felt_result_x), nil
+	} else {
+		return memory.NewMaybeRelocatableFelt(felt_result_y), nil
+	}
+}
+
+func LineSlope(point_a PartialSumB, point_b DoublePointB, prime big.Int) (big.Int, error) {
+	mod_value := new(big.Int).Sub(&point_a.X, &point_b.X)
+	mod_value.Mod(mod_value, &prime)
+
+	if mod_value.Cmp(big.NewInt(0)) == 0 {
+		return big.Int{}, errors.New("is multiple of prime")
+	}
+
+	n := new(big.Int).Sub(&point_a.Y, &point_b.Y)
+	m := new(big.Int).Sub(&point_a.X, &point_b.X)
+
+	z, _ := math_utils.DivMod(n, m, &prime)
+	return *z, nil
+}
+
+func EcAdd(point_a PartialSumB, point_b DoublePointB, prime big.Int) (PartialSumB, error) {
+	m, err := LineSlope(point_a, point_b, prime)
+	if err != nil {
+		return PartialSumB{}, err
+	}
+
+	x := new(big.Int).Mul(&m, &m)
+	x.Sub(x, &point_a.X)
+	x.Sub(x, &point_b.X)
+	x.Mod(x, &prime)
+
+	y := new(big.Int).Mul(&m, new(big.Int).Sub(&point_a.X, x))
+	y.Sub(y, &point_a.Y)
+	y.Mod(y, &prime)
+
+	return PartialSumB{X: *x, Y: *y}, nil
+}
+
+func EcDoubleSlope(point DoublePointB, alpha big.Int, prime big.Int) (big.Int, error) {
+	q := new(big.Int).Mod(&point.Y, &prime)
+	if q == big.NewInt(0) {
+		return big.Int{}, errors.New("is multiple of prime")
+	}
+	n := new(big.Int).Mul(&point.X, &point.X)
+	n.Mul(n, big.NewInt(3))
+	n.Add(n, &alpha)
+
+	m := new(big.Int).Mul(&point.Y, big.NewInt(2))
+	z, _ := math_utils.DivMod(n, m, &prime)
```

Don't ignore the error